br/pkg/lightning/checkpoints/checkpoints.go

```diff
@@ -940,19 +948,32 @@ func (cpdb *MySQLCheckpointsDB) Update(checkpointDiffs map[string]*TableCheckpoi
 type FileCheckpointsDB struct {
 	lock        sync.Mutex // we need to ensure only a thread can access to `checkpoints` at a time
 	checkpoints checkpointspb.CheckpointsModel
+	ctx         context.Context
 	path        string
+	fileName    string
+	exStorage   storage.ExternalStorage
 }
 
-func NewFileCheckpointsDB(path string) *FileCheckpointsDB {
+func NewFileCheckpointsDB(ctx context.Context, path string) (*FileCheckpointsDB, error) {
 	cpdb := &FileCheckpointsDB{
 		path: path,
 		checkpoints: checkpointspb.CheckpointsModel{
 			TaskCheckpoint: &checkpointspb.TaskCheckpointModel{},
 			Checkpoints:    map[string]*checkpointspb.TableCheckpointModel{},
 		},
 	}
+
+	// init ExternalStorage
+	s, fileName, err := createExstorageByCompletePath(ctx, path)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	cpdb.ctx = ctx
+	cpdb.fileName = fileName
+	cpdb.exStorage = s
+
 	// ignore all errors -- file maybe not created yet (and it is fine).
-	content, err := os.ReadFile(path)
+	content, err := cpdb.exStorage.ReadFile(ctx, fileName)
```

Comment:
When underlying storage is s3, the error may not be due to file not exist. Since we have added error to return argument list, I think we can check the error explicitly, and return error when encountered any other errors.

Reply:
may be we can add check before readFile, because error comes from different kinds of storage may different.